kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,591 @@
+
+
+local function stats_init(stats, name)
+    if not stats[name] then
+        stats[name] = {}
+    end
+end
+
+
+local function stats_count(stats, name, key)
+    stats[name][key] = (stats[name][key] or 0) + 1
+end
+
+
+-- For TYPE_LAST: the DNS record from the last successful query
+local valid_types = {
+    [ TYPE_SRV ] = true,
+    [ TYPE_A ] = true,
+    [ TYPE_AAAA ] = true,
+    [ TYPE_CNAME ] = true,
+}
+
+local function insert_last_type(cache, name, qtype)
+    if valid_types[qtype] then
+        cache:set("last:" .. name, { ttl = 0 }, qtype)
+    end
+end
+
+
+local function get_last_type(cache, name)
+    return cache:get("last:" .. name)
+end
+
+
+-- insert hosts into cache
+local function init_hosts(cache, path, preferred_ip_type)
+    local hosts, err = utils.parse_hosts(path)
+    if not hosts then
+        ngx.log(ngx.WARN, "Invalid hosts file: ", err)
+        hosts = {}
+    end
+
+    if not hosts.localhost then
+        hosts.localhost = {
+          ipv4 = "127.0.0.1",
+          ipv6 = "[::1]",
+        }
+    end
+
+    local function insert_answer(name, qtype, address)
+        if not address then
+            return
+        end
+
+        local key = name .. ":" .. qtype
+        local answers = {{
+            name = name,
+            type = qtype,
+            address = address,
+            class = 1,
+            ttl = 0,
+        }}
+        cache:set(key, { ttl = 0 }, answers)
+    end
+
+    for name, address in pairs(hosts) do
+        name = name:lower()
+        if address.ipv4 then
+            insert_answer(name, TYPE_A, address.ipv4)
+            insert_last_type(cache, name, TYPE_A)
+        end
+        if address.ipv6 then
+            insert_answer(name, TYPE_AAAA, address.ipv6)
+            if not address.ipv4 or preferred_ip_type == TYPE_AAAA then
+                insert_last_type(cache, name, TYPE_AAAA)
+            end
+        end
+    end
+
+    return hosts
+end
+
+
+function _M.new(opts)
+    if not opts then
+        return nil, "no options table specified"
+    end
+
+    -- parse resolv.conf
+    local resolv, err = utils.parse_resolv_conf(opts.resolv_conf, enable_ipv6)
+    if not resolv then
+        ngx.log(ngx.WARN, "Invalid resolv.conf: ", err)
+        resolv = { options = {} }
+    end
+
+	-- init the resolver options for lua-resty-dns
+    local nameservers = opts.nameservers or resolv.nameservers
+    if not nameservers or #nameservers == 0 then
+        ngx.log(ngx.WARN, "Invalid configuration, no nameservers specified")
+    end
+
+    local r_opts = {
+        nameservers = nameservers,
+        retrans = opts.retrans or resolv.options.attempts or 5,
+        timeout = opts.timeout or resolv.options.timeout or 2000,   -- ms
+        no_random = opts.no_random or not resolv.options.rotate,
+    }
+
+    -- init the mlcache
+    local lock_timeout = r_opts.timeout / 1000 * r_opts.retrans + 1 -- s
+
+    local cache, err = mlcache.new("dns_cache", "kong_dns_cache", {
```

Comment:
I also strongly advise using appropriate `ttl`, `neg_ttl` configuration values, as well as `shm_miss`. Not splitting the L2 cache between hits and misses opens the attack surface area by allowing an overload of negative cached values (e.g. thousands of worthless domain names) that can slowly evict all cached records out of the L2. It is extremely important to separate the hit/miss caches when using this library in a production, high throughput OpenResty application.

Reply:
Also make sure to properly configure `resty_lock_opts` (e.g. the lock timeout should at least be equal to the DNS query timeout). `shm_set_tries` and `resurrect_ttl` are also worth consideration. `shm_set_tries` matters in case the values stored in the shm vary in size by several orders of magnitude (this is also why it is important to separate `shm` and `shm_miss`, or else memory fragmentation will be very detrimental to the L2 cache effectiveness.

Reply:
I have already utilized the `shm_miss` feature to prevent cache miss attacks. You can refer to this commit: https://github.com/Kong/kong/pull/12305/commits/04c3c37033c0e00760ea090ccb69ee7722dacb21.

Additionally, it's important to note that the original DNS client encounters two cache miss scenarios: one for the `DNS error code=3 NXDOMAIN (name error)` case, and the other for empty records generated when processing the replied records from the nameserver if there are no matching `domain:type` in the result. So returning `nil` from the L3 callback to `mlcache:get()` couldn't distinguish between these two scenarios. In my commit, it only caches the `DNS error code=3 NXDOMAIN (name error)` case as a miss value.